File: apps/formstore/service/models/form_submission.go
// Copyright 2023-2026 Ant Investor Ltd
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package models

import (
	"fmt"

	"github.com/pitabwire/frame/data"
)

// FormSubmissionStatus enumerates submission statuses.
type FormSubmissionStatus string

const (
	SubmissionStatusPending  FormSubmissionStatus = "pending"
	SubmissionStatusComplete FormSubmissionStatus = "complete"
	SubmissionStatusArchived FormSubmissionStatus = "archived"
)

// ValidSubmissionStatuses lists all valid statuses for validation.
var validSubmissionStatuses = map[FormSubmissionStatus]bool{ //nolint:gochecknoglobals // lookup map
	SubmissionStatusPending:  true,
	SubmissionStatusComplete: true,
	SubmissionStatusArchived: true,
}

// ValidateStatus checks whether the status is a known valid value.
func (s FormSubmissionStatus) ValidateStatus() error {
	if !validSubmissionStatuses[s] {
		return fmt.Errorf("invalid submission status: %q", s)
	}

	return nil
}

// FormSubmission stores submitted form data with file references.
type FormSubmission struct {
	data.BaseModel `gorm:"embedded"`

	FormID         string               `gorm:"column:form_id;not null"`
	SubmitterID    string               `gorm:"column:submitter_id"`
	Status         FormSubmissionStatus `gorm:"column:status;not null;default:pending"`
	Data           string               `gorm:"column:data;type:jsonb;not null"`
	FileCount      int                  `gorm:"column:file_count;not null;default:0"`
	IdempotencyKey string               `gorm:"column:idempotency_key"`
	Metadata       string               `gorm:"column:metadata;type:jsonb"`
}

// TableName returns the database table name.
func (FormSubmission) TableName() string {
	return "form_submissions"
}

File: ui_formstore/lib/src/models/form_submission.dart
/// Status of a form submission.
enum SubmissionStatus {
  pending,
  complete,
  archived;

  static SubmissionStatus fromString(String value) {
    return SubmissionStatus.values.firstWhere(
      (s) => s.name == value,
      orElse: () => SubmissionStatus.pending,
    );
  }
}

/// A form submission with its data payload.
class FormSubmission {
  const FormSubmission({
    required this.id,
    required this.formId,
    this.submitterId = '',
    this.status = SubmissionStatus.pending,
    this.data = const {},
    this.fileCount = 0,
    this.idempotencyKey = '',
    this.metadata,
    this.createdAt,
    this.modifiedAt,
  });

  final String id;
  final String formId;
  final String submitterId;
  final SubmissionStatus status;
  final Map<String, dynamic> data;
  final int fileCount;
  final String idempotencyKey;
  final Map<String, dynamic>? metadata;
  final DateTime? createdAt;
  final DateTime? modifiedAt;

  factory FormSubmission.fromJson(Map<String, dynamic> json) {
    return FormSubmission(
      id: json['id'] as String? ?? '',
      formId: json['form_id'] as String? ?? '',
      submitterId: json['submitter_id'] as String? ?? '',
      status: SubmissionStatus.fromString(json['status'] as String? ?? ''),
      data: (json['data'] as Map<String, dynamic>?) ?? const {},
      fileCount: json['file_count'] as int? ?? 0,
      idempotencyKey: json['idempotency_key'] as String? ?? '',
      metadata: json['metadata'] as Map<String, dynamic>?,
      createdAt: json['created_at'] != null
          ? DateTime.tryParse(json['created_at'] as String)
          : null,
      modifiedAt: json['modified_at'] != null
          ? DateTime.tryParse(json['modified_at'] as String)
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (submitterId.isNotEmpty) 'submitter_id': submitterId,
      'data': data,
      if (idempotencyKey.isNotEmpty) 'idempotency_key': idempotencyKey,
      if (metadata != null) 'metadata': metadata,
    };
  }
}
